stan/math/prim/fun/lambert_w.hpp

```diff
@@ -0,0 +1,85 @@
+#ifndef STAN_MATH_PRIM_FUN_LAMBERT_W_HPP
+#define STAN_MATH_PRIM_FUN_LAMBERT_W_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/fun/boost_policy.hpp>
+#include <boost/math/special_functions/lambert_w.hpp>
+
+namespace stan {
+namespace math {
+
+template <typename T, require_arithmetic_t<T>* = nullptr>
+inline auto lambert_w0(const T& x) {
+  return boost::math::lambert_w0(x, boost_policy_t<>());
+}
+
+template <typename T, require_arithmetic_t<T>* = nullptr>
+inline auto lambert_wm1(const T& x) {
+  return boost::math::lambert_wm1(x, boost_policy_t<>());
+}
+
+namespace internal {
+
+/**
+ * Structure to wrap lambert_w0() so it can be vectorized.
+ *
+ * @tparam T type of variable
+ * @param x variable
+ * @return value of the W0 branch of the Lambert W function at x.
+ * @throw std::domain_error if x is smaller than -e^(-1)
```

Comment:
For use in Stan, we generally want to throw an exception so that users can catch errors early.  That's how all of our other distributions work.  About the only place we don't do this is for built-in C++ functions that return NA rather than throwing (like log(-2) or 0 / 0).